### DISCUSSION TITLE
AsyncSession: add and delete methods

### DISCUSSION THREAD

**Discussion by minh-ht** [sqlalchemy/sqlalchemy#8381] [Usage Questions] [answered]
I have a question that I don't seem to find an answer. While using asynchronous SQLAlchemy, I find that AsyncSession.add(obj) is normal function while AsyncSession.delete(obj) is a coroutine and needs await. Why is there such difference?

**Comment by zzzeek** [comment, marked as answer]
the delete() method often needs to run SQL commands in order to apply ORM [cascading](https://docs.sqlalchemy.org/en/14/orm/cascades.html?highlight=cascade#delete) behavior to the object, that it needs to locate related objects that are also to be deleted.   the add() method has no such limitation, even though it also cascades, the objects that it would be affecting are inherently already present in memory.

**Reply by zzzeek** [reply]
https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html?highlight=asyncsession+delete#sqlalchemy.ext.asyncio.AsyncSession.delete

"As this operation may need to cascade along unloaded relationships, it is awaitable to allow for those queries to take place."

**Reply by minh-ht** [reply]
Thank you for the answer! That makes sense. When I thought about delete method, I did not consider its cascading behavior.
